### ISSUE TITLE
Blocks.Math.Mean has invalid sample function

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/Modelica#676]
**Reported by jriel on 5 Jan 2012 16:39 UTC**
In Modelica 3.2, the Blocks.Math.Mean component has the lines

```
  discrete Modelica.SIunits.Time t0 "Start time of simulation";
equation 
  when initial() then
      t0 = time;
  end when;
  when sample(t0+1/f, 1/f) then ...
```

The call to **sample** is invalid.  Section 3.7.3 of the Modelica 3.2 specification states that the start-time argument must be a _parameter expression_. In the usage given, it includes t0, which is not a parameter.  The proper way to handle this is to eliminate t0 (and the accompanying when statement), and replace it with time.start, which is a parameter and evaluates to the start-time of the simulation.

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/676

**Comment by modelica-trac-importer** [comment]
**Modified by stefanv on 5 Jan 2012 16:52 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 6 Jan 2012 08:34 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by ahaumer on 6 Mar 2012 16:32 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by ahaumer on 6 Mar 2012 16:35 UTC**
Would the following be according to the spec / acceptable?

```
protected 
  parameter Modelica.SIunits.Time t0(fixed=false) "Start time of simulation";
  ...
initial equation
  t0 = time;
equation
  ...
```

**Comment by modelica-trac-importer** [comment]
**Modified by ahaumer on 6 Mar 2012 16:36 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 6 Mar 2012 16:43 UTC**
That's a good question. I suspect it is valid, if somewhat convoluted. Why not just use time.start inside the call to sample as Joe suggested? Then you can get rid of t0 and the initial equation section entirely.

**Comment by modelica-trac-importer** [comment]
**Comment by jriel on 6 Mar 2012 16:58 UTC**
I agree with Stefan's comment---it appears as though it would work, but that can be done more directly with

```
   when sample(time.start + 1/f, 1/f) then
      ...
   end when;
```

If you do want to assign a t0 parameter, simpler would be
```
protected 
    parameter Modelica.SIunits.Time t0 = time.start "Start time of simulation";
```
and elide the initial equation section.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 6 Mar 2012 17:15 UTC**
It is not allowed to access the start value of a variable, so time.start is not correct Modelica (start and other attributes can only be accessed in a modifier).

I think the only (cumbersome) way is the solution from Toni (define t0 as parameter and initialize it in an intial equation section).

**Comment by modelica-trac-importer** [comment]
**Comment by jriel on 6 Mar 2012 17:24 UTC**
Replying to [comment:8 otter]:
> It is not allowed to access the start value of a variable, so time.start is not correct Modelica (start and other attributes can only be accessed in a modifier).
> 
> I think the only (cumbersome) way is the solution from Toni (define t0 as parameter and initialize it in an intial equation section).

Could you point out where that is specified?  For a specific counter-example, see the top of page 199 of Fritzson:
```
  parameter Real event_time = time.start + 20;
```

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 6 Mar 2012 17:40 UTC**
You are right, I quickly checked and in the Modelica 3.2 specification it seems not defined how attributes can be accessed (I had in my mind that this was defined in section 4.8 - Predefined Types).

It is clear that there must be restrictions since, e.g., otherwise in a function call it would be also possible to access every attribute of every input argument. Furthermore, it would be possible to set attributes in the equation section (like min, max, nominal, ...), or it would be possible to access every attribute in the equation section (so generated C-code would have to support it as well).

It seems that this needs to be properly defined.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 6 Mar 2012 17:41 UTC**
Sorry forgot to login. The last comment was from me, Martin.

**Comment by modelica-trac-importer** [comment]
**Comment by jriel on 6 Mar 2012 17:50 UTC**
Replying to [comment:11 otter]:
Thanks for the followup.  My assumption is that **time** is special, so it is allowed.  Being able to reference an abitrary x.start value would be interesting.  As you point out, the actual restrictions should be specified somewhere.

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 6 Mar 2012 18:34 UTC**
I agree that `time` could be special, since the issue of referring to its attributes within functions cannot come up, since referring to `time` in functions is already disallowed.

I don't see why it being "possible to access every attribute in the equation section" is a problem. These attributes are, after all, simply parameters, and there is no restriction on accessing general parameters in the equation section. However, I agree that being able to define (i.e. solve for) the values of attributes in the initial equation section is problematic.

**Comment by modelica-trac-importer** [comment]
**Comment by ahaumer on 6 Mar 2012 19:52 UTC**
fixed by 6b5dd764c1337d6a1ef6149b82837709acd758df, 10b38c96c201e31d4c9e92999008cab64603fb04, fd2eed53f5627934fd8ec96c3747e20c672b6baf, 9ecd0ce05df6f5d6c667e890e6a4b2884e26107c, e57ec9ba2139281c5a93dfa00550a8551fa9179a, 3b7a26f6cfd1d72d2b10a7e25da589d6fe59ab28, e841be83008ad3a2643330fcff923ff01a9027e3, 093e78be2afb6431842e63f0183e5b6d89a0d835 (with initial equation)
